### ISSUE TITLE
range end index out of range for slice of length...

### ISSUE THREAD

**Issue by siosphere** [swc-project/swc#6467]
### Describe the bug

This is an intermittent bug with @swc/jest and/or @swc/core. 

We are running our test suite with turbo repo, utilizing @swc/jest, and intermittently we get a failure like this:
```
failed to handle: range end index 7551504 out of range for slice of length 6467552
@ssi/document-landing:test: 
@ssi/document-landing:test:       at Compiler.transformSync (../../node_modules/@swc/core/index.js:241:29)
@ssi/document-landing:test:       at transformSync (../../node_modules/@swc/core/index.js:348:21)
@ssi/document-landing:test:       at Object.process (../../node_modules/@swc/jest/index.js:73:45)
```
Retrying the ci/cd job and it will succeed just fine. 

This does not happen very often, but ~1/50 runs. That error message appears to be from rust, which is pointing me more towards it being an @swc issue, than a jest specific issue.

### Config

```json
{
    "jsc": {
        "parser": {
            "syntax": "typescript",
            "tsx": true
        }
    },
    "module": {
        "type": "commonjs"
    }
}
```

### Expected behavior

Test suites should not intermittently fail with no changes

### Actual behavior

Fails intermittently

### Version

1.2.128

### Additional context

"@swc/cli": "^0.1.57",
"@swc/core": "^1.2.128",
"@swc/jest": "^0.2.15",

**Comment by kwonoj** [comment]
We wish to dig this down, but as you may be aware it is extremely non trivial to debug these kind of intermittent issue without having an isolated repro codes.

We'll try to look into time to time, but probably will be blocked by no repro.

**Comment by siosphere** [comment]
> We wish to dig this down, but as you may be aware it is extremely non trivial to debug these kind of intermittent issue without having an isolated repro codes.
> 
> We'll try to look into time to time, but probably will be blocked by no repro.

Understandable. I'm wondering if there is a way to get more debug output out of swc. I could setup our CI/CD pipeline to run with more verbose output if it is available. (I'm not familiar enough with rust to go poking around and adding debug statements, but if a verbose option doesn't currently exist I may attempt that)

**Comment by kwonoj** [comment]
It is not easy, but if you could run this with debug build it maybe helpful. But it means

- build swc manually to produce debug build
- adjust CI setup to replace npm installed binary to use above debug build
- bite huge performance drops (debug build is much slower) until repro happens

**Comment by siosphere** [comment]
Since it is happening daily, it is worth it for me to go down that path. 

I'll take a look at the build directions for swc and get a custom debug build setup, I'll update this issue with any findings.

Thanks,

**Comment by siosphere** [comment]
Some additional output:
```
@coursehero/html-viewer:test: thread '<unnamed>' panicked at 'range end index 7551504 out of range for slice of length 1626080', /usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/wasmer-engine-universal-2.3.0/src/artifact.rs:102:38
@coursehero/html-viewer:test: note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Comment by kdy1** [comment]
Can you try `RUST_BACKTRACE=1`?

**Comment by kwonoj** [comment]
>  no activity for a month

🤔 this seems not correct

**Comment by kwonoj** [comment]
@kdy1 some ideas? Issue is opened a wee ago, and last comment was 6 days ago so there's physically no way this issue can be a `month` old without activity.

**Comment by kdy1** [comment]
The ment js wrong
Just remove need more info label

**Comment by siosphere** [comment]
Finally was able to get some more time to look at this, and got an error with backtrace available:
```thread '<unnamed>' panicked at 'range end index 7551504 out of range for slice of length 6303712', /usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/wasmer-engine-universal-2.3.0/src/artifact.rs:102:38
Stack backtrace:
0: <unknown>
1: <unknown>
2: <unknown>
3: <unknown>
4: <unknown>
5: _ZN6v8impl12_GLOBAL__N_123FunctionCallbackWrapper6InvokeERKN2v820FunctionCallbackInfoINS2_5ValueEEE
6: _ZN2v88internal12_GLOBAL__N_119HandleApiCallHelperILb0EEENS0_11MaybeHandleINS0_6ObjectEEEPNS0_7IsolateENS0_6HandleINS0_10HeapObjectEEESA_NS8_INS0_20FunctionTemplateInfoEEENS8_IS4_EENS0_16BuiltinArgumentsE
7: _ZN2v88internal21Builtin_HandleApiCallEiPmPNS0_7IsolateE
8: Builtins_CEntry_Return1_DontSaveFPRegs_ArgvOnStack_BuiltinExit```

re-running the job and it passes just fine. And it is not always the same test or even the same package that errors

**Comment by dgreif** [comment]
Dropping in to link https://github.com/swc-project/plugins/issues/42, which I believe has a similar root cause. I was not able to get the debug info previously (thanks for doing the hard work on that @siosphere!), but I'm seeing a little more detail on my end after updating to the latest swc:

```
thread '<unnamed>' panicked at 'range end index 6294120 out of range for slice of length 2158560', /usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/wasmer-engine-universal-2.3.0/src/artifact.rs:102:38
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Note the `wasmer-engine-universal-2.3.0/src/artifact.rs:102:38` was not part of the original stack trace for my issue or this one, but seems to perfectly match with what @siosphere reported above.  Not positive, but I'm guessing it points to [this line from the wasmer 2.3.0 release](https://github.com/wasmerio/wasmer/blob/2.3.0/lib/engine/src/artifact.rs#L102)? I don't see anything obviously related there, but hopefully @kdy1 or @kwonoj can spot something 🤞

**Comment by kdy1** [comment]
https://github.com/wasmerio/wasmer/blob/2.3.0/lib/engine-universal/src/artifact.rs#L102 is the correct link

**Comment by kdy1** [comment]
And yes, this happens if `metadata_len` is too large

**Comment by dgreif** [comment]
@kdy1 thanks for verifying the source. Is this something the SWC team will be able to look into, or should we upstream the issue to wasmer?

**Comment by kdy1** [comment]
It's not a something we can look into

**Comment by dgreif** [comment]
I had originally been under the impression this was only happening on some specific docker images (much older images), and that the errors weren't present on `ubuntu:latest` in GitHub Actions. We recently bumped our actions runners to much larger instances with 16 cores, and are starting to see these errors in that environment as well. From this, it seems that running many concurrent instances of swc + jest _in parallel_ leads to this issue.

**Comment by HashemKhalifa** [comment]
I can relate to that. I tried using the GitHub Actions runner and everything went smoothly. However, when using our self-hosted runner in parallel mode, we encountered some issues. To mitigate the problem, I tried running it with the --runInBand option, and while the issue still occurred, it became less frequent.

**Comment by ezpuzz** [comment]
Also getting this issue intermittently, sometimes resolved by deleting `.swc` directory to re-download the loadable plugin

**Comment by abejfehr** [comment]
> It's not a something we can look into

Can you elaborate why not?

**Comment by samar-1601** [comment]
Has anyone got any resolution for this issue?
@kdy1 Do we have any updates/workaround on this. I have a very large monorepo which fails the CI pipes intermittently, giving the following error:
![image](https://github.com/user-attachments/assets/ae0b9cdc-3fd9-4f46-bacf-07dff3a30b10)

**Comment by HashemKhalifa** [comment]
@samar-1601 which jest version are you using and `react-testing-library`? I 

It was a memory leak that happened while using an older version of jest and `react-testing-library` that you had to use a different way of handling the events.

**Comment by HashemKhalifa** [comment]
here is what I updated and adjust my tests to match the new updates.
```
    "@swc-node/register": "1.8.0",
    "@swc/cli": "0.1.62",
    "@swc/core": "1.3.62",
    "@swc/helpers": "0.5.1",
    "@swc/jest": "0.2.26",
    "@swc/plugin-jest": "^1.5.67",
    "@swc/plugin-loadable-components": "0.3.67",
    "@swc/plugin-styled-components": "1.5.67",
    "jest": "^29.7.0",
    "jest-axe": "^8.0.0",
    "jest-canvas-mock": "^2.5.2",
    "jest-each": "^29.7.0",
    "jest-environment-jsdom": "^29.7.0",
    "jest-localstorage-mock": "^2.4.26",
    "jest-styled-components": "^7.1.1",
    "jest-transform-stub": "^2.0.0",
    "jest-watch-typeahead": "^2.2.2",
    "jest_workaround": "^0.76.0",
    "jsdom": "^22.1.0",
    "jsonc-eslint-parser": "^2.1.0",
      "@testing-library/dom": "^9.3.3",
    "@testing-library/jest-dom": "^6.1.4",
    "@testing-library/react": "^14.0.0",
    "@testing-library/user-event": "^14.4.3",
```

**Comment by samar-1601** [comment]
@HashemKhalifa We are using the following versions.
```javascript
"jest": "26.6.3",
"jest_workaround": "^0.72.6",
``` 
> It was a memory leak that happened while using an older version of jest and react-testing-library that you had to use a different way of handling the events.

Is this related in any way to these node+jest version issues.
- https://github.com/jestjs/jest/issues/11956
- https://github.com/jestjs/jest/issues/7874
We have tried using various combinations of maxWorkers=(some percentage), -w=(some number of threads). But none of the memory management steps solves the problem in a concrete manner i.e. this `range end index` error keeps on coming intermittently.

I have found possible solutions(if the problem is related to jest+node version issues) which includes upgrading node to 21.1 and using workerIdleMemoryLimit(mentioned in the [official jest documentation](https://jestjs.io/docs/next/configuration#workeridlememorylimit-numberstring)). 

The problem is I am stuck in a 2 way here, cause the solutions
- Upgrading node + jest
- Upgrading swc
Both will take a huge effort considering the size of our monorepo and currently, we have our pipes failing intermittently on a daily basis.

**Comment by HashemKhalifa** [comment]
I feel you, been in your shoes, and that's correct with the memory leak issues you mentioned, I had to upgrade both, since then all good, and currently planning to move away from jest to vitest as well.

**Comment by samar-1601** [comment]
@HashemKhalifa Did you use the `workerIdleMemoryLimit` feature to restrict the memory as well? Cause in our case, upgrading node to 21.1 as mentioned in https://github.com/jestjs/jest/issues/11956 didn't solve our problem, instead it became a recurring issue. Hoping the jest upgrade to 29.7 and `workerIdleMemoryLimit` solves our problem. 

Will update the thread with our findings.

**Comment by HashemKhalifa** [comment]
I have tried `workerIdleMemoryLimit `, but it didn't work as I expected, which react version are you using?

**Comment by samar-1601** [comment]
@HashemKhalifa React version is `18.2.0`
So if `workerIdleMemoryLimit` didn't work, correct me if I'm wrong, just upgrading node and jest solved the issue?

**Comment by HashemKhalifa** [comment]
@samar-1601 no, was one of the reasons that caused memory leaks, as well in our test cases there were so many tests that I had to adjust to fix the leaks, some of which had a relation with `react-testing-library` packages. like `@testing-library/jest-dom` and `@testing-library/user-event`

I hope that answers your question

**Comment by kdy1** [comment]
Removing the milestone as there's no repro.

**Comment by samar-1601** [comment]
@HashemKhalifa Another observation, using flags `no-compilation-cache` and `expose-gc` reduce the memory consumption of the jest test cases.
Here is the result of a sample run over one of our modules(`jest v26.4`):
```
- node v16.10
    - Without `no-compilation-cache`, no `expose-gc` → 900+ MB
    - With the above, 580 MB (259 sec)
- node v18.2 (clearly has leaks)
    - Without `no-compilation-cache`, no `expose-gc` → 1300+ MB
    - With the above → 3000+ MB (then I manually stopped it)
- node v21.1.0
    - Without `no-compilation-cache`, no `expose-gc` → 1200+ MB (150 sec)
    - With the above →  around 220-230 MB avg (163 sec)

**Comment by issacgerges** [comment]
We too see this consistently. Are there any other diagnostic steps you can recommend? We're running these builds on C7s  (xl I believe) so it's hard to believe the issue is related to memory exhaustion. Any info you can give about the error that might help us understand the issue? Is this reading compiled code from a cache? Or reading the compilation response?

It's a long shot but considering the worker model of jest is it possible that multiple workers are requesting the same file/module to be compiled at the same time?

**Comment by samar-1601** [comment]
Yeah, this may be related to any issue when jest runs with multiple workers cause the error doesn't come when we use `--runInBand`(`--runInBand` is synonymous to `-w=1` i.e 1 worker thread). But the problem is that we can't go ahead with one worker only because of the time taken for the tests to execute.
For instance, a process which takes 8-9s to run with `maxWorkers=50%`, 22s for `-w=4`, takes 98-100s with `--runInband`.

**Comment by dgreif** [comment]
`--runInBand` is still the only solution we have found which consistently avoids this error. We use `nx` for build parallelization and caching in our monorepo, which allows us to get _most_ of the benefits provided by multiple Jest workers, but we do have a few _big_ packages which still run slowly with `--runInBand` and delay the overall test run from finishing. I wish I had more insight for y'all, but that's all I've got 😅

**Comment by samar-1601** [comment]
@HashemKhalifa can you guide a bit as to how you detected leaks and then fixed them. 
For us, currently using `--no-compilation-cache` seems to fix the memory pile up issue.
Here is a sample for the same:

| Flag | Without | With |
|------|--------|-------|
| `--no-compilation-cache` |  ![image](https://github.com/user-attachments/assets/bdd381b6-b51e-4b9f-9699-7c2b7f3ac9d4)   |  ![image](https://github.com/user-attachments/assets/dbb6dc77-6af1-44bb-9dcb-42e2cae2e179)    |

**Comment by HashemKhalifa** [comment]
I can share my findings while working on this issue last year maybe it could help:

Updating Node reduced the leaks but didn't solve the problem completely  https://github.com/jestjs/jest/issues/11956 
https://issues.chromium.org/issues/42202158

Updating Jest still didn't solve the problem entirely but reduced how often it happened. 

I'm not sure if it's related to SWC, but it definitely exacerbates the issue in Jest, I was hoping `--workerIdleMemoryLimit` would solve the issue but it didn't have any effect.

As @dgreif mentioned, there's only one option, which is runInBand, because Jest instances with memory leaks are not able to shut down and keep running forever until SWC complains.

Our solution:

Run tests in-band (sequentially) within each Jest instance.
Keep parallelization at a higher level using NX, running multiple Jest instances concurrently.
This approach isolates each test instance, preventing memory leaks from spreading, while still leveraging parallel execution for improved overall performance.

**Comment by PartMan7** [comment]
Chiming in to mention that this is still a major issue for us; several of our (~3000-large) tests in Jest end up failing with this. It's not a consistently reproducible issue either, but it's been impacting us for several years now...

```
Summary of all failing tests
FAIL TEST.spec.tsx
  ● Test suite failed to run
    failed to handle: range end index 4323496 out of range for slice of length 2506720
      at Compiler.transformSync (../../node_modules/@swc/core/index.js:241:29)
      at transformSync (../../node_modules/@swc/core/index.js:348:21)
      at Object.process (../../node_modules/@swc/jest/index.js:73:45)
      at ScriptTransformer.transformSource (../../node_modules/@jest/transform/build/ScriptTransformer.js:464:35)
      at ScriptTransformer._transformAndBuildScript (../../node_modules/@jest/transform/build/ScriptTransformer.js:569:40)
      at ScriptTransformer.transform (../../node_modules/@jest/transform/build/ScriptTransformer.js:607:25)
```

Like others have mentioned, this does get 'fixed' through runInBand, but that's not a solution since it triples the runtime (and is practically impossible to use on our larger repo with ~150K tests).

**Comment by kdy1** [comment]
I dug into this and the failure strongly looks like a race while multiple Jest processes populate the same compiled Wasm plugin cache file. As a temporary workaround, each worker can use a separate `jsc.experimental.cacheRoot`.

It is safer to set this in a small transformer wrapper loaded inside the worker, rather than directly in `jest.config.js` (which may be evaluated by the parent process):

```js
// swc-jest-transformer.cjs
const path = require("node:path");
const { createTransformer } = require("@swc/jest");

const workerId = process.env.JEST_WORKER_ID ?? "in-band";

module.exports = createTransformer({
  // Keep the rest of your existing SWC options here.
  jsc: {
    // Keep the rest of your existing jsc options here.
    experimental: {
      // Keep your existing plugins/options here.
      cacheRoot: path.join(
        process.cwd(),
        ".swc",
        `jest-${process.ppid}-${workerId}`,
      ),
    },
  },
});
```

Then point Jest at the wrapper:

```js
module.exports = {
  transform: {
    "^.+\\.[jt]sx?$": "<rootDir>/swc-jest-transformer.cjs",
  },
};
```

`JEST_WORKER_ID` alone may collide when Turbo/Nx starts multiple independent Jest instances, because each instance numbers its workers from 1. Including the parent PID (or another per-Jest-run/package identifier) avoids that collision.

This trades cache reuse and some startup performance for isolation, but should avoid workers observing a partially-written shared plugin cache. `--runInBand` remains the simpler workaround if the performance cost is acceptable.

**Comment by kdy1** [comment]
But I'll fix this in the next version
